bootstrap/bootstrap.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// bootstrap implements a drop-in-replacement for jenkins/bootstrap.py
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	flag "github.com/spf13/pflag"
+)
+
+// Constant Keys for known environment variables and URLs
+const (
+	BuildEnv          string = "BUILD_NUMBER"
+	BootstrapEnv      string = "BOOTSTRAP_MIGRATION"
+	CloudSDKEnv       string = "CLOUDSDK_CONFIG"
+	GCEKeyEnv         string = "JENKINS_GCE_SSH_PRIVATE_KEY_FILE"
+	Gubernator        string = "https://k8s-gubernator.appspot.com/build"
+	HomeEnv           string = "HOME"
+	JenkinsHomeEnv    string = "JENKINS_HOME"
+	JobEnv            string = "JOB_NAME"
+	NodeEnv           string = "NODE_NAME"
+	ServiceAccountEnv string = "GOOGLE_APPLICATION_CREDENTIALS"
+	WorkspaceEnv      string = "WORKSPACE"
+	GCSArtifactsEnv   string = "GCS_ARTIFACTS_DIR"
+)
+
+// Args contains all of the (parsed) command line arguments for bootstrap
+// NOTE: Repo should be further parsed by ParseRepos
+type Args struct {
+	Root           string
+	Timeout        int32
+	Repo           []string
+	Bare           bool
+	Job            string
+	Upload         string
+	ServiceAccount string
+	SSH            string
+	GitCache       string
+	Clean          bool
+}
+
+// ParseArgs parses the command line to an Args instance
+// arguments should be generally be os.Args[1:]
+func ParseArgs(arguments []string) (*Args, error) {
+	args := &Args{}
+	flags := flag.NewFlagSet("bootstrap", flag.ContinueOnError)
+	// used to mimic required=true
+	requiredFlags := []string{"job"}
```

Comment:
check and error out if the flag is empty should be enough?

Reply:
possibly. I ported 100% semantics from the python arg parsing. it seemed smarter to complain specifically for not providing the --job flag. not setting a valid job is probably a problem handled somewhere further up the chain?

Reply:
most of other places simply do, and probably easier to understand
```
if *foo = "" {
    error
}
```

Reply:
Ok, Let's do both though so we can give more specific errors. I feel like if you don't set `--job` at all you aren't familiar with bootstrap, vs if you manage to set `--job=""` you probably have a bug in a calling script or something.